File: src/org/cerion/symcalc/expression/function/list/IdentityMatrix.java
package org.cerion.symcalc.expression.function.list;

import org.cerion.symcalc.exception.ValidationException;
import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.NumberExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class IdentityMatrix extends FunctionExpr {

    public IdentityMatrix(Expr... e) {
        super(FunctionType.IDENTITY_MATRIX,e);
    }

    @Override
    protected Expr evaluate() {
        int n = getInteger(0).intValue();
        ListExpr result = new ListExpr();

        for(int i = 0; i < n; i++) {
            ListExpr sublist = new ListExpr();
            for(int j = 0; j < n; j++)
                sublist.add(i == j ? IntegerNum.ONE : IntegerNum.ZERO);

            result.add(sublist);
        }

        return result;
    }

    @Override
    public void validate() throws ValidationException {
        validateParameterCount(1);
        validateParameterType(0, ExprType.NUMBER);
        validateNumberType(0, NumberExpr.INTEGER);

        int n = getInteger(0).intValue();
        if (n <= 0)
            throw new ValidationException("Number must be a positive integer");
    }
}

File: src/org/cerion/symcalc/expression/function/list/IdentityMatrix.kt
package org.cerion.symcalc.expression.function.list

import org.cerion.symcalc.exception.ValidationException
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.FunctionExpr
import org.cerion.symcalc.expression.ListExpr
import org.cerion.symcalc.expression.NumberExpr
import org.cerion.symcalc.expression.number.IntegerNum

class IdentityMatrix(vararg e: Expr) : FunctionExpr(FunctionExpr.FunctionType.IDENTITY_MATRIX, *e) {

    override fun evaluate(): Expr {
        val n = getInteger(0).intValue()
        val result = ListExpr()

        for (i in 0 until n) {
            val sublist = ListExpr()
            for (j in 0 until n)
                sublist.add(if (i == j) IntegerNum.ONE else IntegerNum.ZERO)

            result.add(sublist)
        }

        return result
    }

    @Throws(ValidationException::class)
    override fun validate() {
        validateParameterCount(1)
        validateParameterType(0, Expr.ExprType.NUMBER)
        validateNumberType(0, NumberExpr.INTEGER)

        val n = getInteger(0).intValue()
        if (n <= 0)
            throw ValidationException("Number must be a positive integer")
    }
}
